File: packages/nest-crud-request/src/lib/relation-builder.ts
import { RelationObject, RelationObjectValue, RelationOptions } from './types';


export class RelationBuilder {

    private relations: RelationObject = {};

    constructor(relations?: RelationOptions | string) {
        if (relations) {
            const relationOptions = typeof relations === 'string' ? JSON.parse(relations) : relations;
            this.setRelations(relationOptions);
        }
    }

    setRelations(relations: RelationOptions): this {
        if (relations) {
            if (Array.isArray(relations)) {
                relations.forEach(relation => {
                    this.add(relation);
                });
            } else if (typeof relations === 'string') {
                this.add(relations);
            } else {
                this.relations = relations;
            }
        }
        return this;
    }

    clear(): this {
        this.relations = {};
        return this;
    }

    /**
     * Add (or replace) a relation. Two call styles, both optional after the name:
     *
     * ```ts
     * builder.add('posts');                                   // join, all columns
     * builder.add('posts', ['id', 'title']);                  // pick columns
     * builder.add('posts', ['id'], { status: 'published' });  // + relation-scoped where
     * builder.add('posts', ['id'], undefined, 'inner');       // + inner join
     * builder.add('posts', { select: ['id'], joinType: 'inner' }); // object config
     * ```
     */
    add(
        relation: string,
        select?: string[] | RelationObjectValue,
        where?: Record<string, any>,
        joinType?: 'left' | 'inner',
    ): this {
        // Object-config form: add('posts', { select, where, joinType })
        if (select && !Array.isArray(select) && typeof select === 'object') {
            const config = select as RelationObjectValue;
            this.relations[relation] = Object.keys(config).length ? { ...config } : true;
            return this;
        }

        // Positional form
        const config: RelationObjectValue = {};
        if (select) {
            config.select = select as string[];
        }
        if (where) {
            config.where = where;
        }
        if (joinType) {
            config.joinType = joinType;
        }
        this.relations[relation] = Object.keys(config).length ? config : true;
        return this;
    }

    remove(relation: string): this {
        delete this.relations[relation];
        return this;
    }

    hasRelations(): boolean {
        return Object.keys(this.relations).length > 0;
    }

    toObject(): RelationObject {
        return this.relations;
    }

    toJson(): string {
        return JSON.stringify(this.relations);
    }

}

File: clients/flutter/nest_crud_request/lib/src/relation_builder.dart
import 'dart:convert';

/// Builds the `relations` object, mirroring the JS `RelationBuilder`.
class RelationBuilder {
  Map<String, dynamic> _relations = {};

  RelationBuilder([dynamic relations]) {
    if (relations != null) setRelations(relations);
  }

  RelationBuilder setRelations(dynamic relations) {
    if (relations is List) {
      for (final r in relations) {
        add(r as String);
      }
    } else if (relations is String) {
      _relations = Map<String, dynamic>.from(jsonDecode(relations) as Map);
    } else if (relations is Map) {
      _relations = Map<String, dynamic>.from(relations);
    }
    return this;
  }

  RelationBuilder clear() {
    _relations = {};
    return this;
  }

  /// Add (or replace) a relation. With no options it's a plain join (`true`);
  /// otherwise the per-relation `{ select, where, joinType }` config is emitted.
  RelationBuilder add(
    String relation, {
    List<String>? select,
    Map<String, dynamic>? where,
    String? joinType,
  }) {
    if (select == null && where == null && joinType == null) {
      _relations[relation] = true;
    } else {
      final config = <String, dynamic>{};
      if (select != null) config['select'] = select;
      if (where != null) config['where'] = where;
      if (joinType != null) config['joinType'] = joinType;
      _relations[relation] = config;
    }
    return this;
  }

  RelationBuilder remove(String relation) {
    _relations.remove(relation);
    return this;
  }

  bool hasRelations() => _relations.isNotEmpty;
  Map<String, dynamic> toObject() => _relations;
  String toJson() => jsonEncode(_relations);
}
